Implement basic per-CPU / CPU-local storage

kevinaboos opened the pull request:
* Not yet used anywhere
* Only supported on x86_64

Reminder: remove temp tests and log statements before merging

kevinaboos committed:
[no ci] WIP temp saving cpu_local crate

kevinaboos committed:
Merge branch 'theseus_main' into cpu_locals

kevinaboos committed:
[no ci] WIP, cpu locals mostly work. STill needs improvement

kevinaboos committed:
Finish CPU-local variables / per-CPU storage

tsoutsman commented on kernel/ap_start/src/lib.rs:
```diff
@@ -102,9 +102,10 @@ pub fn kstart_ap(
         nmi_flags,
     ).unwrap();
-
+    
```
Whitespace.

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+// NOTE: These fields must be kept in sync with `cpu_local::FixedCpuLocal`.
+impl FixedCpuLocal {
+    const SELF_PTR_OFFSET: usize = 0;
+    pub const CPU_ID:                       Self = Self { offset: 8,  size: 4, align: 4 };
+    pub const PREEMPTION_COUNT:             Self = Self { offset: 12, size: 1, align: 1 };
+    pub const TASK_SWITCH_PREEMPTION_GUARD: Self = Self { offset: 16, size: 8, align: 4 };
+    pub const DROP_AFTER_TASK_SWITCH:       Self = Self { offset: 24, size: 8, align: 8 };
```
Could we somehow test these offsets in `per_cpu` using `const_assert`?

kevinaboos replied:
i was trying to figure out a way to do that, but I don't think it's possible given the cyclic dependency issues it would cause. If you have another idea though, I'm open to it and would love to have these be more thoroughly tested and checked for errors, especially because this design is brittle towards change.

kevinaboos replied:
nvm, i misread what you suggested. Done now.

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+    ///    of this new data image, making it "currently active" and accessible.
+    fn new() -> Result<CpuLocalDataImage, &'static str> {
+        // 1. Allocate a single page to store each CPU's local data.
+        let mut mp = memory::create_mapping(1, PteFlags::new().writable(true).valid(true))?;
```
Minor thing, but asserting that we fit into a single page i.e.
```rust
const_assert_eq!(DROP_AFTER_TASK_SWITCH.offset + DROP_AFTER_TASK_SWITCH.size, memory::PAGE_SIZE)
```

kevinaboos replied:
Ah yes, good call. An even better solution would be to pass in the size of the data structure from `per_cpu`, which ensures we'll have enough memory.

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+        unsafe {
+            #[cfg(target_arch = "x86_64")]
+            asm!(
+                concat!("mov {}, gs:[0]"), // the SELF_PTR_OFFSET is 0
```
Is there a reason for `concat!`?

kevinaboos replied:
leftover relic of old changes, thanks for catching.

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+    ///
+    /// Thus, the caller must guarantee that the type `T` is correct for the
+    /// given `FixedCpuLocal`.
+    pub const unsafe fn new_fixed(
```
When do we use this function?

kevinaboos replied:
we don't use it yet, but it would be used like so:
```rust
// in `kernel/preemption`:
static PREEMPTION_COUNT: CpuLocal<..., PreemptionCount> = unsafe { CpuLocal::new_fixed(FixedCpuLocal::PREEMPTION_COUNT) };

// in hold_preemption, increase the count:
let prev_val = PREEMPTION_COUNT.with(|count| count.fetch_add(1, ...));

// in PreemptionGuard::drop(), decrease the count:
let prev_val = PREEMPTION_COUNT.with(|count| count.fetch_sub(1, ...));
```

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+
+    // Run the given initializer function on the per-CPU data.
+    let new_data_image = CpuLocal::<{FixedCpuLocal::SELF_PTR_OFFSET}, P>(PhantomData);
+    new_data_image.with_mut(|per_cpu_data_mut| {
```
Converting the address to a `&mut P` (in `with_mut`) even if we don't use `per_cpu_data_mut` is undefined behaviour as the memory is uninitialised.

kevinaboos replied:
ya, basically i was looking at clever ways to avoid `core::ptr::write()` here, but i don't think there is a good way. You're right.

tsoutsman commented on kernel/per_cpu/src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+    /// A pointer to the start of this struct in memory, similar to a TLS self pointer.
+    /// This has a different initial value for each CPU's data image, of course.
+    ///
+    /// We use this to allow writes to this entire structure (for initialization),
+    /// and also to allow faster access to large fields herein () accelerate accesses to large items
+    self_ptr: usize,
```
Could you elaborate on how this field allows faster access to large fields because this field is causing the problems with uninitialised memory in `cpu_local`?

kevinaboos replied:
basically it avoids having to load the entire field before modifying it. I forgot to link to this discussion in the comments, but I meant to add this in: https://github.com/rust-osdev/x86_64/pull/257#issuecomment-849514649

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+impl<const OFFSET: usize, T: Copy> CpuLocal<OFFSET, T> {
+    /// Returns a copy of this `CpuLocal`'s inner value of type `T`.
+    ///
+    /// This is a convenience functiononly available for types where `T: Copy`.
```
```suggestion
    /// This is a convenience function only available for types where `T: Copy`.
```

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+///
+/// Note that this struct doesn't contain an instance of the type `T`,
+/// and dropping it has no effect.
+pub struct CpuLocal<const OFFSET: usize, T>(PhantomData<*mut T>);
```
If we use the enum representation above, I *think* this could be defined as:
```rust
unsafe trait Field: Sized {
    // Idk if this would work.
    const SIZE_CHECK: () = const_assert_eq!(core::mem::size_of::<Self>(), Self::field().size());
    const ALIGN_CHECK: () = const_assert_eq!(core::mem::align_of::<Self>(), Self::field().align());

    const fn field() -> CpuLocalField;
}

pub struct CpuLocal<T: Field>(PhantomData<*mut T>);
```

the offset would be calculated using `T::field().offset()`.

kevinaboos replied:
ah ok, so now I remember why i moved away from an enum+trait based design -- you can't have a `const fn` in traits yet.

kevinaboos replied:
if the rust PR for "enum variants as types" ever gets merged in, then we could improve this, but for now I hacked around that using enum variants that contain empty structs for each field.
EDIT: nvm, this didn't work.

kevinaboos replied:
Actually, we need the user of `CpuLocal` to be able to specify the type `T`, and at least with the design I had in mind, that user wouldn't be able to impl our private trait for their type `T` ... unless we defined and impl'd that trait inside `per_cpu`, which would introduce another cyclic dep problem. hmm

kevinaboos replied:
fixed in latest changeset

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+use x86_64::{registers::model_specific::GsBase, VirtAddr};
+
+/// Info use to obtain a reference to a CPU-local variable; see [`CpuLocal::new()`].
+pub struct FixedCpuLocal {
```
I think the name and implementation are quite confusing. Maybe something like this would be better:
```rust
pub enum CpuLocalField {
    SelfPointer,
    CpuId,
    PreemptionCount,
    TaskSwitchPreemptionGuard,
    DropAfterTaskSwitch,
}

impl CpuLocalField {
    const fn offset(&self) -> usize {
        match self {
            SelfPointer => 0,
            CpuId => 8,
            PreemptionCount => 12,
            TaskSwitchPreemptionGuard => 16,
            DropAfterTaskSwitch => 24,
    }

    // const fn size

    // const fn align
}

tsoutsman replied:
See comment below for more details.

kevinaboos replied:
yea i like the enum approach and actually had that as my first design attempt. couldn't figure out how to do the const asserts easily, but I'll try it with your trait idea above!

tsoutsman reviewed (commented):
I've proposed a slightly different API that would decrease the amount of unsafe *syncing* we have to do between `PerCpuData` and `FixedCpuLocal`.

Just to make sure we're on the same page `gs` is a register that isn't used in regular code and so we can use it for a different purpose right?

kevinaboos commented:
> Just to make sure we're on the same page `gs` is a register that isn't used in regular code and so we can use it for a different purpose right?

Correct, the GS segment register is unused and is frequently used for CPU-local storage on x86. It's similar to how the FS segment register is used for thread-local storage.

kevinaboos committed:
address all concerns except trying new enum/trait-based design

kevinaboos committed:
Add assertions in `per_cpu` for offset definitions in `cpu_local`

kevinaboos commented:
@tsoutsman I attempted a few different ways to use traits in the `CpuLocal` type, with the goal being to make the RAII declaration of a CPU-local variable as simple as possible. It's pretty hard to make the type signature simple, but I was able to get it down to a single generic type parameter (which itself has one type parameter). I didn't push it to this branch because I don't love it, but take a look here: https://github.com/theseus-os/Theseus/commit/3b24b1daddb215a4e352027a49115b8396579bbf.  At the very least, it removes/hides the `const OFFSET` parameter from `CpuLocal`, which I do think was a bit of a flawed/leaky design. This design lets the caller do something like:
```rust
static PREEMPTION_COUNT: CpuLocal<CpuLocalPreemptionCount<PreemptionCount>> = unsafe { CpuLocal::new() };
```

--------------

The only other option I was able to get to work was to parameterize the `CpuLocalVariable` trait with a type, like so:
```rust
pub trait CpuLocalVariable<T: Sized> { ... }
```
which works, but you then need to have both `T` and `V` parameters for `CpuLocal`:
```rust
pub struct CpuLocal<T: Sized, V: CpuLocalVariable<T>> {
    _typ: PhantomData<T>,
    _var: PhantomData<V<T>>,
}

impl<T: Sized, V: CpuLocalVariable<T>> CpuLocal<T, V> {
    pub const unsafe fn new(_variable_marker: V) -> Self {
        Self { _typ: PhantomData, _var: PhantomData }
    }
```
This is clearer but ergonomically worse, as you have to declare/instantiate it like so:
```rust
static PREEMPTION_COUNT: CpuLocal<PreemptionCount, CpuLocalVariable<PreemptionCount>> = unsafe { CpuLocal::new(CpuLocalPreemptionCount) };
```

-------------
If you have any ideas on how to further simplify this, please do let me know.

tsoutsman committed:
Enum API

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman commented:
Apologies force push was because I pushed to the wrong remote repo.

Here's a quick enum implementation I made:
https://github.com/theseus-os/Theseus/commit/b0ce230b862b6eb22a96cc2fc716214e58b2dc99

It removes size and alignment from `cpu_local`, removes `OFFSET` const parameter, prevents `CpuLocal` from being instantiated with arbitrary offsets and types. The one bad thing is that `per_cpu` now needs to define a bunch of wrapper types so that it can implement `Field` for them.

tsoutsman commented:
right but that introduces a circular dependency hmm

we could have the Field impls in the crated where the type is defined e.g. cpu for CpuId

kevinaboos commented:
Thanks, that enum impl is similar to what I tried, but the addition of the unsafe trait with its associated const does help. I was trying to avoid an unsafe trait -- i wonder if there's a clever way to prevent other crates from implementing the `Field` trait.

> The one bad thing is that `per_cpu` now needs to define a bunch of wrapper types so that it can implement `Field` for them.

This is fine, but we should impl Deref+DerefMut on those wrappers so the crates that use `cpu_local` don't have to be aware of said newtype wrappers.

> we could have the Field impls in the crates where the type is defined, e.g. cpu for CpuId

Yeah, i don't really have any other ideas for workarounds here, so I think this is fine. But we should expose it via a macro like `define_cpu_local_type!(<type>)` such that it's not easy/recommended to directly impl `Field`. This ties into trying to prevent someone from implementing `Field`, but I do realize that macros wouldn't prevent that.

I suppose you could also define a proc_macro that would implement hidden traits for a cpu-local type, but that might be too complex right now. My future idea for TLS-style CPU locals incorporates proc_macros for custom attributes, but we can discuss that later.

---------

Finally, you left this comment, which I agree with, but yes that's one of the many reasons why `CpuLocal::new*` was unsafe -- the type provided had to match the offset. 
> // NOTE: The fact that this previously compiled in `cpu_local` is
>        // indicative of an unsafe API, as the offsets (and types) are
>        // completely arbitrary. There's nothing stopping us from trying to
>        // access CpuLocal::<16, String> which would be undefined behaviour.

kevinaboos commented:
> Here's a quick enum implementation I made: [b0ce230](https://github.com/theseus-os/Theseus/commit/b0ce230b862b6eb22a96cc2fc716214e58b2dc99)
> 
> It removes size and alignment from `cpu_local`, removes `OFFSET` const parameter, prevents `CpuLocal` from being instantiated with arbitrary offsets and types. The one bad thing is that `per_cpu` now needs to define a bunch of wrapper types so that it can implement `Field` for them.

Thanks, I have merged/included that commit into this PR branch, and am going from there.

> we could have the Field impls in the crated where the type is defined e.g. cpu for CpuId

Working on this now. I've also addressed some naming issues.

--------

Will re-request a review from you when done.

kevinaboos committed:
Use trait-based design for CpuLocal

kevinaboos committed:
Merge branch 'theseus_main' into cpu_locals

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,333 @@
+    ($TypeWrapper:ident, $InnerType:ty, $PerCpuField:expr) => {
+        #[repr(transparent)]
+        struct $TypeWrapper($InnerType);
+        unsafe impl $crate::CpuLocalField for $TypeWrapper {
+            const FIELD: $crate::PerCpuField = $PerCpuField;
+        }
```
I don't like the fact that we have a macro to implement an unsafe trait.

The trait isn't marked unsafe to satisfy the compiler; it's marked unsafe to warn the user that they must uphold certain invariants. The following definition would cause undefined behaviour but would never use the unsafe keyword:
```rust
define_per_cpu_field!(CpuId, cpu::CpuId, PerCpuField::PreemptionCount);
```

There isn't much point in marking it unsafe if we provide a macro that avoids the `unsafe` keyword completely.

kevinaboos replied:
That's fair, agreed.

tsoutsman commented on kernel/cpu_local/src/lib.rs:
```diff
@@ -0,0 +1,333 @@
+///
+/// This macro implements the [`CpuLocalField`]
+#[macro_export]
+macro_rules! define_per_cpu_field {
```
Can we move this macro into `per_cpu`? No other crate should be using this macro.

kevinaboos replied:
Agreed, will do.

kevinaboos replied:
Actually I'll remove it entirely.

tsoutsman commented on kernel/per_cpu/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+    }
+}
+
+define_per_cpu_field!(CpuId, cpu::CpuId, PerCpuField::CpuId);
+define_per_cpu_field!(PreemptionCount, preemption::PreemptionCount, PerCpuField::PreemptionCount);
+define_per_cpu_field!(TaskSwitchPreemptionGuard, Option<preemption::PreemptionGuard>, PerCpuField::TaskSwitchPreemptionGuard);
+define_per_cpu_field!(DropAfterTaskSwitch, Option<TaskRef>, PerCpuField::DropAfterTaskSwitch);
```
AFAICT, this creates a circular dependency that `per_cpu` was trying to avoid in the first place.

`preemption` will not be able to access the CPU local `PreemptionCount` as that would lead to a circular dependency (`per_cpu` depends on `preemption` depends on `per_cpu`). The trait has to be implemented in the crate where the type is defined (e.g. in `cpu` or `preemption`).

kevinaboos replied:
Yeah, that's true.

`preemption` is a special case, given that `cpu_local` itself must depend on `preemption` because its `CpuLocal::with()` accessors use `PreemptionGuard` and `hold_preemption()`. So in the next PR when we start using per-CPU preemption counts, I will have to merge `cpu_local` and `preemption`. That's a known issue, so no worries there.

I'll work on this.
```rust

```

kevinaboos committed:
* impl newtype wrappers for Cpu-local compatibility in the crates
  that define the inner type being wrapped.
* CpuId is a special case.
* We can't yet impl CpuLocalField for PreemptionCount, as that
  requires merging the `cpu_local` and `preemption` crates.

kevinaboos commented:
Made the changes, thanks for the suggestion. Though not included in this changeset, I confirmed it's possible to instantiate a CPU-local task type, i.e., this works:
```rust
static TASK_SWITCH_PREEMPTION_GUARD: cpu_local::CpuLocal<TaskSwitchPreemptionGuard> = cpu_local::CpuLocal::new(PerCpuField::TaskSwitchPreemptionGuard);
pub fn test_me() -> Option<PreemptionGuard> {
    TASK_SWITCH_PREEMPTION_GUARD.with_mut(|t| t.0.take())
}
```

see [my latest commit msg](https://github.com/theseus-os/Theseus/pull/917/commits/5d64794e646fdef4f8bb5f75b8e997ea0b19f321) for other details

kevinaboos committed:
fuck off clippy

tsoutsman commented on kernel/cpu_local/src/lib.rs:
````diff
@@ -0,0 +1,309 @@
+    ///
+    /// The type `T: CpuLocalField` must be specified with the turbofish operator:
+    /// ```rust,no_run
+    /// static CPU_ID: CpuLocal<CpuId> = CpuId::new(PerCpuField::CpuId);
````
```suggestion
    /// static CPU_ID: CpuLocal<CpuLocalCpuId> = CpuId::new(PerCpuField::CpuId);
````

tsoutsman reviewed (approved):
One small doc change.

kevinaboos committed:
doc fix

Co-authored-by: Klim Tsoutsman <klimusha@gmail.com>

kevinaboos committed:
remove code for testing cpu locals

kevinaboos committed:
remove log msg

kevinaboos committed:
fix aarch64 build
